[web] Update package:web to ">=0.5.1 <2.0.0"

ditman opened the pull request:
This PR updates packages that depend on `package:web` to `web: ">=0.5.1 <2.0.0"`.

## Issues

* Fixes: https://github.com/flutter/flutter/issues/151981

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

ditman commented:
(First push to see how CI likes this, expecting a bunch of errors!)

ditman commented:
```
Because google_identity_services_web >=0.3.1+1 depends on web ^0.5.1 and google_identity_services_web >=0.3.1 <0.3.1+1 depends on web ^0.5.0, google_identity_services_web >=0.3.1 requires web ^0.5.0.
So, because google_sign_in_web depends on both google_identity_services_web ^0.3.1 and web ^1.0.0, version solving failed.
```

```
Because google_maps_flutter_web depends on google_maps ^7.1.0 which depends on web ^0.5.0, web ^0.5.0 is required.
So, because google_maps_flutter_web depends on web ^1.0.0, version solving failed.
```


* `google_identity_services_web` needs to be published first.
* We also need an update of `google_maps` to `web: ^1.0.0` (maybe in the 8 branch?)

yjbanov reviewed (approved):
![lgtm](https://user-images.githubusercontent.com/211513/32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d.png)

Rexios80 commented:
@ditman There is some cleanup to do that won't show up as build errors since a lot of the import statements are like this:
```dart
import 'package:web/web.dart' as web;
```
which will hide duplicate type errors

`package:web` 1.0.0 added a lot of types that currently exist in custom extensions. For example in `video_player_web`:

```dart
import 'package:web/web.dart' as web;

/// Adds a "disablePictureInPicture" setter to [web.HTMLVideoElement]s.
extension NonStandardSettersOnVideoElement on web.HTMLVideoElement {
  // TODO(srujzs): This will be added in `package:web` 0.6.0. Remove this helper
  // once it's available.
  external set disablePictureInPicture(bool disabled);
}

/// Adds a "disableRemotePlayback" and "controlsList" setters to [web.HTMLMediaElement]s.
extension NonStandardSettersOnMediaElement on web.HTMLMediaElement {
  // TODO(srujzs): This will be added in `package:web` 0.6.0. Remove this helper
  // once it's available.
  external set disableRemotePlayback(bool disabled);
  external set controlsList(String? controlsList);
}
```

stuartmorgan commented:
For any packages without any code changes (which is currently all of them, but I'm not sure how preliminary that is), consider expanding the range to include 1.x instead of requiring 1.x. According to pub there are about 150 packages depending on `web`, including a bunch of popular ones. The more packages do a gradual transition allowing 0.5 *or* 1.x for a little while instead of going straight to requiring 1.x, the fewer people will have resolver errors that prevent them from using the latest versions of all of their packages.

Not a requirement by any means, but something to consider.

ditman commented:
> package:web 1.0.0 added a lot of types, functions, fields, etc that currently exist in custom extensions. For example in video_player_web:

@Rexios80 you're right, I'll eventually get to those as we fix the more core dependencies. There's only a few "package_web_tweaks" in the different packages which should be "easy to remove".

> will hide duplicate type errors

@Rexios80 we have fairly strict warnings-as-errors in the build; I think those errors will eventually show up once the "all_packages" app is able to resolve to web:^1.0.0. Or do you mean those will fail at run-time or some time after publishing?

> For any packages without any code changes (which is currently all of them, but I'm not sure how preliminary that is), consider expanding the range to include 1.x instead of requiring 1.x.

Good call @stuartmorgan, I'll try to expand the untouched packages to a range, as long as the code is compatible.

Rexios80 commented:
> we have fairly strict warnings-as-errors in the build; I think those errors will eventually show up once the "all_packages" app is able to resolve to web:^1.0.0. Or do you mean those will fail at run-time or some time after publishing?

I'm not sure they will show up as warnings/errors since technically nothing will be wrong. The extensions will just shadow the real types (if the web import is named) and fields (always). They should still build fine, there will just be superfluous code.

ditman commented:
> I'm not sure they will show up as warnings/errors since technically nothing will be wrong. The extensions will just shadow the real types (if the web import is named) and fields (always). They should still build fine, there will just be superfluous code.

@Rexios80 Ah yes, I think this is fine (for a while). IMO it's very nice that we can write this type of "bridge" code across major versions for a change! 😭

We can (and should) remove the `package_web_tweaks` files once we bump the dependency to `^1.0.0` only (and not the range). For example, we could use this PR to clean up the small tweaks that `google_maps_flutter_web` has to `package:web`, if we want to make it web:1.0.0 only.

Rexios80 commented:
> As @stuartmorgan said, the packages that don't require actual code changes to go to web:^1.0.0, should support the >=0.5.1 <2.0.0 range until they must be moved to ^1.0.0 (because of us needing some feature from the package). It seems that the tweaks needed to support the web range are pretty small anyways! 🤞

In the past I would be opposed to maintaining a version range on a dependency since it's really easy to accidentally break backwards compatibility. However `pana` now checks packages after running `pub downgrade` so this is no longer an issue!

stuartmorgan commented:
> In the past I would be opposed to maintaining a version range on a dependency since it's really easy to accidentally break backwards compatibility. However `pana` now checks packages after running `pub downgrade` so this is no longer an issue!

We've been doing that in our CI, as part of presubmit checks [for two years](https://github.com/flutter/plugins/pull/6207). The pana check was actually added because I recommended it based on our setup.

kevmoo commented:
@ditman – do we just need to rebase this on latest?

ditman commented:
@kevmoo I had to do two things on top of that:

* bump a patch version (as suggested by stuart in a comment above)
* make web a `>=0.5.1 <2.0.0` range

I expect some things to fail (I didn't run the full analyze in my local machine), but everything should be fixable with some extension method that lets me do what I was doing in the old version of the code.

As Rexios mentioned above, we're going to need an issue when we move away from web:0.5.1 to clean-up all those little shims and actually adopt web:1.0.0 fully.

ditman commented:
Also, to enable bracketing web: 0.5.1 to 2.0.0, I can't make dart 3.4 the minimal version... If so, web 0.5.1 would never resolve (and I could as well just say: web 1.0.0).

I'm removing dart: ^3.4 and flutter: >=3.22, and leaving those at 3.3 and 3.19, so web 0.5.1 has a chance of resolving.

ditman commented:
I've created this issue for the cleanup effort:

* https://github.com/flutter/flutter/issues/152657

PTAL @Rexios80 / @stuartmorgan!

This PR migrates all the web packages to the following dependencies:

* dart: `^3.3`
* flutter: `>=3.19`
* web: `>=0.5.1 <2.0.0`

ditman commented:
(This should trigger all tests?)

Rexios80 commented on packages/file_selector/file_selector_web/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.9.4+2
+
+* Supports `web: ">=0.5.1 <2.0.0"`.
+* Updates SDK version to Flutter `^3.19.0`.
```
The Flutter version doesn't appear to have changed. I think we can remove this line.

ditman replied:
Yes! This is a typo from `0.9.4`: https://github.com/flutter/packages/commit/b21dce5f4964adf1e6c90225377a00f65652c428#diff-a1970a9bbbd9cd744445641d16604a32f71cb82413a2dc02d216eacd5101616d (that I'm about to retcon)

Rexios80 commented on packages/google_sign_in/google_sign_in_web/lib/src/flexible_size_html_element_view.dart:
```diff
@@ -78,6 +78,8 @@ class _FlexHtmlElementView extends State<FlexHtmlElementView> {
     final web.DOMRectReadOnly rect = resizes.toDart.last.contentRect;
     if (rect.width > 0 && rect.height > 0) {
+      // The following ignore can be removed once the package leaves web:0.5.1 behind.
+      //ignore:noop_primitive_operations
```
Add a space

`// ignore`

Rexios80 replied:
Actually there should be space after the colon as well:

`// ignore: noop_primitive_operations`

ditman replied:
Yep, I used the "quick fix" from vscode, and it told me it should be this:

```
// ignore: noop_primitive_operations
```

Rexios80 commented on packages/web_benchmarks/pubspec.yaml:
```diff
@@ -2,7 +2,7 @@ name: web_benchmarks
 repository: https://github.com/flutter/packages/tree/main/packages/web_benchmarks
 issue_tracker: https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+label%3A%22p%3A+web_benchmarks%22
-version: 2.0.0
+version: 2.0.1
```
Since there is a code change in this package, should this be `2.1.0`?

ditman replied:
Nah, the code changes are so minimal that I think a patch version should be enough.

Rexios80 reviewed (commented):
Just some small things I noticed. Otherwise LGTM!

Rexios80 commented:
@ditman Should we ask for a test exemption on Discord?

ditman commented:
> @ditman Should we ask for a test exemption on Discord?

@Rexios80 it's almost 8PM PT, so it's probably a little bit late (or too early) for Discord. 

@stuartmorgan has test exemption powers and can also answer [your question about the `web_benchmarks` version](https://github.com/flutter/packages/pull/7202#discussion_r1699297597); I'm sure this'll be ready to `autosubmit` early tomorrow!

stuartmorgan commented on packages/google_identity_services_web/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.3.1+4
+
+* Downgrades SDK version to Dart `^3.3.0`.
```
I would say "Restores support for Dart 3.3."; "downgrade" sounds like something is getting worse.

ditman replied:
Yep, better. Changed!

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_web/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.5.9+2
+
+* Downgrades SDK version to Dart `^3.3.0`. Flutter `^3.19.0`.
```
Same.

ditman replied:
Ditto.

stuartmorgan commented on packages/google_sign_in/google_sign_in_web/example/pubspec.yaml:
```diff
@@ -2,8 +2,8 @@ name: google_sign_in_web_integration_tests
-  sdk: ^3.3.0
-  flutter: ">=3.19.0"
+  sdk: ^3.4.0
+  flutter: ">=3.22.0"
```
Why are a bunch of examples being given tighter SDK constraints than the library?

ditman replied:
No reason, I just didn't roll back the dependencies there, in case somebody uses the examples as copy-paste, at least they get "the latest".

ditman replied:
(Rolled back)

stuartmorgan commented on packages/google_sign_in/google_sign_in_web/lib/src/flexible_size_html_element_view.dart:
```diff
@@ -78,6 +78,8 @@ class _FlexHtmlElementView extends State<FlexHtmlElementView> {
   ) {
     final web.DOMRectReadOnly rect = resizes.toDart.last.contentRect;
     if (rect.width > 0 && rect.height > 0) {
+      // The following ignore can be removed once the package leaves web:0.5.1 behind.
```
This should be a TODO.

ditman replied:
Made a TODO, and linked to the cleanup issue.

stuartmorgan commented on packages/cross_file/CHANGELOG.md:
```diff
@@ -1,6 +1,6 @@
+## 0.3.4+2
 
-* Updates minimum supported SDK version to Flutter 3.16/Dart 3.2.
+* Supports `web: ">=0.5.1 <2.0.0"`.
```
FWIW I would say "Adds support for `web` 1.x`." for all of these, rather than putting the full constraint details in the CHANGELOG.

kevmoo replied:
I kinda like the full constraint...just makes it SUPER clear what's going on. But 🤷

ditman replied:
Ret-conned the full constraint into ^1.0.0 (in practice I think it's easier to understand, maybe?)

stuartmorgan commented:
test-exempt: code refactor with no semantic change

kevmoo commented:
`auto-submit`?

Rexios80 commented:
@kevmoo There are some small tweaks left

ditman commented:
OK, I've addressed all of the comments. It seems that if CI is happy, we're all happy, so adding `autosubmit` to this PR (will keep an eye on flakes just in case)

ditman commented:
This is now **published**, thanks!

* https://docs.flutter.dev/platform-integration/web/embedding-flutter-web
* https://docs.flutter.dev/platform-integration/web/web-content-in-flutter
